src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -0,0 +1,694 @@
+        /// <param name="name">The name of the paramter to add or update.</param>
+        /// <param name="value">The value of the parameter to add or update.</param>
+        /// <remarks>
+        /// If <paramref name="value"/> is <c>null</c>, the parameter will be removed if it exists in the URI.
+        /// Otherwise, it will be added or updated.
+        /// </remarks>
+        public static string UriWithQueryParameter(this NavigationManager navigationManager, string name, string? value)
+        {
+            if (navigationManager is null)
+            {
+                throw new ArgumentNullException(nameof(navigationManager));
+            }
+
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException(EmptyQueryParameterExceptionMessage, nameof(name));
+            }
+
+            var uri = navigationManager.Uri;
+
+            return value is null
+                ? UriWithoutQueryParameter(uri, name)
+                : UriWithQueryParameterCore(uri, name, value);
+        }
+
+        /// <summary>
+        /// Returns a <see cref="string"/> equal to <see cref="NavigationManager.Uri"/> except with a single parameter
+        /// updated with the provided <paramref name="values"/>.
+        /// </summary>
+        /// <param name="navigationManager">The <see cref="NavigationManager"/>.</param>
+        /// <param name="name">The name of the paramter to add or update.</param>
+        /// <param name="values">The value of the parameter to add or update.</param>
+        /// <remarks>
+        /// Any <c>null</c> entries in <paramref name="values"/> will be skipped. Existing querystring parameters not in
+        /// <paramref name="values"/> will be removed from the querystring in the returned URI.
+        /// </remarks>
+        public static string UriWithQueryParameter<TValue>(this NavigationManager navigationManager, string name, IEnumerable<TValue> values)
+        {
+            if (navigationManager is null)
+            {
+                throw new ArgumentNullException(nameof(navigationManager));
+            }
+
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException(EmptyQueryParameterExceptionMessage, nameof(name));
+            }
+
+            var uri = navigationManager.Uri;
+
+            if (!TryRebuildExistingQueryFromUri(uri, out var existingQueryStringEnumerable, out var newQueryStringBuilder))
+            {
+                return UriWithAppendedQueryParameters(uri, name, values);
+            }
+
+            var formatter = GetFormatterFromParameterValueType(typeof(TValue));
+            var encodedName = Uri.EscapeDataString(name).AsSpan();
+            var valueEnumerator = values.GetEnumerator();
+            var hasNextValue = valueEnumerator.MoveNext();
+
+            foreach (var pair in existingQueryStringEnumerable)
+            {
+                if (pair.EncodedName.Span.SequenceEqual(encodedName))
```

Comment:
I think you can use `
```suggestion
                if (pair.EncodedName.Span.Equals(encodedName, StringComparison.OrdinalIgnoreCase))
```